[PP] Refactor test_schedule_multiproc

H-Huang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #159591
* #158978
* __->__ #158780

This refactors the pipelining schedule tests since a lot of them have the same repeated code of:
1. Create pipelined model and reference model
2. Run reference model and pipelined model
3. compare gradients

So this refactors those parts above into helper methods and reduces ~300 LOC. Also adds a better gradient check to resolve flakiness (fixes https://github.com/pytorch/pytorch/issues/154408).

cc @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta

H-Huang committed:
Refactor test_schedule_multiproc

[ghstack-poisoned]

wconstab commented on test/distributed/pipelining/test_schedule_multiproc.py:
```diff
@@ -60,38 +60,197 @@ def backend_str(cls) -> str:
+        if n_layers is None:
+            n_layers = self.world_size
+        if model_class is None:
+            model_class = MultiMLP
```
nit: make model_class=MultiMLP in the fn decl?

wconstab commented on test/distributed/pipelining/test_schedule_multiproc.py:
```diff
@@ -60,38 +60,197 @@ def backend_str(cls) -> str:
+
+        return ref_out, ref_loss
+
+    def _run_schedule_steps(self, schedule, x, target=None, num_iterations=2, **kwargs):
```
i'm a little dubious of the value of extracting the runner loop as a helper.  The other helpers all seem great to me.  is it more clear if the test just loops over calling the model and the ref?  I don't have a strong objection here.

wconstab commented on test/distributed/pipelining/test_schedule_multiproc.py:
```diff
@@ -60,38 +60,197 @@ def backend_str(cls) -> str:
+
+        # First, try the standard strict comparison
+        try:
+            torch.testing.assert_close(grad1, grad2, rtol=rtol, atol=atol)
```
what's the point of even doing this assertion?  We'll never know if we hit this case or fell through to the flexible case, so to me this amounts to making every test use the relative path.  Is that what you wanted, or do you want some tests to be strict and some to be flexible?  

Do we know why some tests were failing with the strict check?

H-Huang replied:
I think I understand it better now. Since the `loss_fn` is `MSELoss(reduction="sum")` then increasing the samples in the batch will lead to larger loss values

The losses were quite large, for example in `test_grad_with_manual` and 2 microbatches I am seeing:

```python
losses = [tensor(65603.1562, device='cuda:1', grad_fn=<MseLossBackward0>), tensor(65587.8594, device='cuda:1', grad_fn=<MseLossBackward0>)] 
ref_loss = tensor(131191., device='cuda:1', grad_fn=<MseLossBackward0>)
```

Already see we precision loss here because `sum(losses) = 131191.0156` and ref_loss is just `131191`. On top of this the floating point arithmetic isn't associative so `(grad_from_loss1) + (grad_from_loss2)` ≠ `grad_from_(loss1 + loss2)`

This explains why reducing the batch size 256 -> 64 helps. I also tried `MSELoss(reduction="mean")` and the gradients are much closer as well.

wconstab reviewed (commented):
could you split into 2 PRs (1) the refactor part, which i'll stamp (2) the change in behavior of the tests, which i'm wondering about

H-Huang committed:
Update on "[PP] Refactor test_schedule_multiproc"


This refactors the pipelining schedule tests since a lot of them have the same repeated code of:
1. Create pipelined model and reference model
2. Run reference model and pipelined model
3. compare gradients

So this refactors those parts above into helper methods and reduces ~300 LOC. Also adds a better gradient check to resolve flakiness (fixes https://github.com/pytorch/pytorch/issues/154408).

cc awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta 

[ghstack-poisoned]

H-Huang committed:
Update on "[PP] Refactor test_schedule_multiproc"


This refactors the pipelining schedule tests since a lot of them have the same repeated code of:
1. Create pipelined model and reference model
2. Run reference model and pipelined model
3. compare gradients

So this refactors those parts above into helper methods and reduces ~300 LOC. Also adds a better gradient check to resolve flakiness (fixes https://github.com/pytorch/pytorch/issues/154408).

cc awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta 

[ghstack-poisoned]

H-Huang committed:
Update on "[PP] Refactor test_schedule_multiproc"


This refactors the pipelining schedule tests since a lot of them have the same repeated code of:
1. Create pipelined model and reference model
2. Run reference model and pipelined model
3. compare gradients

So this refactors those parts above into helper methods and reduces ~300 LOC. Also adds a better gradient check to resolve flakiness (fixes https://github.com/pytorch/pytorch/issues/154408).

cc awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta 

[ghstack-poisoned]

wconstab reviewed (approved):
thanks for cleaning this up!
